src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -16,118 +29,198 @@
-     *
-     * @return An immutable reference to the Parameter registry
-     */
-    static const std::vector<std::unique_ptr<Parameter<T>>>& getRegistry()
-    {
-        return Parameter<T>::getMutableRegistry();
-    }
-
-    /**
-     * Registers (adds) a Parameter to the registry. Since the unique pointer is moved
-     * into the registry, the pointer may not be accessed by the caller after this
-     * function has been called.
-     *
-     * @param parameter A unique pointer to the Parameter to add. This pointer may not
-     * be accessed by the caller after this function has been called.
-     */
-    static void registerParameter(std::unique_ptr<Parameter<T>> parameter)
-    {
-        Parameter<T>::getMutableRegistry().emplace_back(std::move(parameter));
-    }
-
-    /**
-     * Updates all the Parameters of type T with the latest values from the ROS
-     * Parameter Server
-     */
-    static void updateAllParametersFromROSParameterServer()
-    {
-        for (const auto& p : Parameter<T>::getRegistry())
+    public:
+        /**
+         * Constructs a new Parameter
+         *
+         * @param parameter_name The name of the parameter used by dynamic_reconfigure
+         * @param default_value The default value for this parameter
+         */
+        explicit Parameter<T>(const std::string& parameter_name, T default_value)
+        {
+            this->name_ = parameter_name;
+            this->value_ = default_value;
+
+            Parameter<T>::registerParameter(std::make_unique<Parameter<T>>(*this));
+        }
+
+        /**
+         * Returns the global path in the ROS parameter server where this parameter is stored
+         *
+         * @return the global path in the ROS parameter sever where this parameter is stored
+         */
+        const std::string getROSParameterPath() const
+        {
+            return kParameterNs + "/" + name();
+        }
+
+        /**
+         * Returns the value of this parameter
+         *
+         * @return the value of this parameter
+         */
+        const T value() const
+        {
+            return value_;
+        }
+        /**
+         * Returns the name of this parameter
+         *
+         * @return the name of this parameter
+         */
+        const std::string name() const
+        {
+            return name_;
+        }
+
+        /**
+         * Updates the value of this Parameter with the value from the ROS
+         * Parameter Server
+         */
+        void updateValueFromROSParameterServer()
+        {
+            ros::param::get(getROSParameterPath(), value_);
+        }
+
+        /**
+         * Sets the parameter value in the ROS Parameter Server to the new value.
+         *
+         * @param new_value The new value to be set
+         */
+        void setValueInROSParameterServer(T new_value)
+        {
+	    //dynamic reconfigure takes control of setting params //ros::param::set(getROSParameterPath(), new_value);
```

Comment:
you can set the parameter through code, but it will not be picked up on the GUI, (behind the scenes it will use the new value), not sure how to go about this, remove entirely or leave this in? @MathewMacDougall

Reply:
Oh darn. This might be a bit of a pain, since we were aiming to do parameter management from our side of things (ie. control things via the `Parameter` class....).

The solution here might just be to save the configuration values to a file and then let ROS manage them all. Not too sure though. @MathewMacDougall - your thoughts?

Reply:
We can still use parameters and control things in our code (set/get `Parameter`), but if we allow that param to be externally set (through the dynamic reconfigure gui), the value being set in the gui would conflict with the value being changed around in the code, so would we ever set values to the params available for configuration through the gui?

Reply:
Historically, our code does not set dynamic parameters, only the GUI does. I think in this case it's fine to treat Parameters as "read-only" from the POV of the code and only have them modified externally (through a GUI by the user)

Reply:
using the dynamic reconfigure service actually will update the gui and the param server at the same time, so ill change that function to make that service call, so we could do that if we wanted